mycroft/util/file_utils.py

```diff
@@ -0,0 +1,238 @@
+        LOG.info('Low diskspace detected, cleaning cache')
+        # calculate how many bytes we need to delete
+        bytes_needed = (min_free_percent - percent_free) / 100.0 * space.total
+        bytes_needed = int(bytes_needed + 1.0)
+
+        # get all entries in the directory w/ stats
+        entries = (os.path.join(directory, fn) for fn in os.listdir(directory))
+        entries = ((os.stat(path), path) for path in entries)
+
+        # leave only regular files, insert modification date
+        entries = ((stat[ST_MTIME], stat[ST_SIZE], path)
+                   for stat, path in entries if S_ISREG(stat[ST_MODE]))
+
+        # delete files with oldest modification date until space is freed
+        space_freed = 0
+        for moddate, fsize, path in sorted(entries):
+            try:
+                os.remove(path)
+                space_freed += fsize
+            except Exception:
+                pass
+
+            if space_freed > bytes_needed:
+                return  # deleted enough!
+
+
+def get_cache_directory(domain=None):
+    """Get a directory for caching data.
+
+    This directory can be used to hold temporary caches of data to
+    speed up performance.  This directory will likely be part of a
+    small RAM disk and may be cleared at any time.  So code that
+    uses these cached files must be able to fallback and regenerate
+    the file.
+
+    Args:
+        domain (str): The cache domain.  Basically just a subdirectory.
+
+    Return:
+        str: a path to the directory where you can cache data
+    """
+    config = mycroft.configuration.Configuration.get()
+    dir = config.get("cache_path")
+    if not dir:
+        # If not defined, use /tmp/mycroft/cache
+        dir = os.path.join(tempfile.gettempdir(), "mycroft", "cache")
+    return ensure_directory_exists(dir, domain)
+
+
+def ensure_directory_exists(directory, domain=None):
+    """ Create a directory and give access rights to all
+
+    Args:
+        domain (str): The IPC domain.  Basically a subdirectory to prevent
+            overlapping signal filenames.
+
+    Returns:
+        str: a path to the directory
+    """
+    if domain:
+        directory = os.path.join(directory, domain)
+
+    # Expand and normalize the path
+    directory = os.path.normpath(directory)
+    directory = os.path.expanduser(directory)
+
+    if not os.path.isdir(directory):
+        try:
+            save = os.umask(0)
+            os.makedirs(directory, 0o777)  # give everyone rights to r/w here
+        except OSError:
+            LOG.warning("Failed to create: " + directory)
+            pass
+        finally:
+            os.umask(save)
+
+    return directory
+
+
+def create_file(filename):
+    """ Create the file filename and create any directories needed
+
+        Args:
+            filename: Path to the file to be created
+    """
+    try:
+        os.makedirs(os.path.dirname(filename))
+    except OSError:
+        pass
```

Comment:
ensure_directory_exists would need to be changed a bit since by default it sets the rights to 777 and that's not something that is always what you want.

Reply:
This was updated to use the `ensure_directory_exist()` with an extra parameter.